apis/v1alpha2/httproute_types.go

```diff
@@ -62,44 +62,29 @@ type HTTPRouteSpec struct {
 	// 1. IPs are not allowed.
 	// 2. The `:` delimiter is not respected because ports are not allowed.
 	//
-	// Incoming requests are matched against the hostnames before the
-	// HTTPRoute rules. If no hostname is specified, traffic is routed
-	// based on the HTTPRouteRules.
-	//
-	// Requests will be matched against the Host field in the following order:
-	//
-	// 1. If Hostname is precise, the request matches this rule if
-	//    the HTTP Host header is equal to the Hostname.
-	// 2. If Hostname is a wildcard, then the request matches this rule if
-	//    the HTTP Host header is to equal to the suffix
-	//    (removing the first label) of the wildcard rule.
-	// 3. If Hostname is unspecified, empty, or `*`, then any request will match
-	//    this route.
-	//
-	// If a hostname is specified by the Listener that the HTTPRoute is bound
-	// to, at least one hostname specified here must match the Listener specified
-	// hostname as per the rules above. Other hostnames will not affect processing
-	// of the route in that case.
-	//
-	// If no hostname is specified by the Listener, then that value will be treated
-	// as '*', match any hostname, and so any hostname on this Route will match.
-	//
-	// If all hostnames do not match, then the HTTPRoute is not admitted, and
-	// the implementation must raise an 'Admitted' Condition with a status of
-	// `false` for that Listener.
-	//
-	// Examples:
-	// - A Listener with unspecified, empty, or `*` values for Hostname matches
-	//   any HTTPRoute hostname.
-	// - A HTTPRoute with unspecified, empty, or `*` values for Hostname matches
-	//   any Listener hostname.
-	// - A Listener with `test.foo.com` as the hostname matches *only*
-	//   `test.foo.com` or `*.foo.com`. Any other hostnames present must be ignored.
-	// - A Listener with `*.foo.com` as hostname, all hostnames in the HTTPRoute
-	//   must have any single label where the star is, and the rest of the hostname
-	//   must match exactly. So, `test.foo.com`, `*.foo.com` or `blog.foo.com` match.
-	//   `test.blog.foo.com`, `test.bar.com`, or `bar.com` do not. Hostnames that do
-	//   not match will be ignored.
+	// If a hostname is specified by both the Listener and HTTPRoute, their
+	// must be at least one intersecting hostname for the HTTPRoute to be
+	// attached to the Listener. For example:
+	//
+	// * A Listener with `test.example.com` as the hostname matches HTTPRoutes
+	//   that have either not specified any hostnames, or have specified at
+	//   least one of `test.example.com` or `*.example.com`.
```

Comment:
```suggestion
	//   least one of `test.example.com` or `*.example.com` or `*`.
```

Reply:
This is actually something that came out of API review, I can't remember which of the reviewers suggested it, but the recommendation was to remove "*" as a possible value. An empty or unspecified value here would continue to be interpreted at "*", but there was no point in differentiating between empty and "*" since they were interpreted the same way. Let me add that to the PR description.